### PR TITLE
Normal priority for low sends

### PR DESCRIPTION
**Pull request by jimleroyer**
# Summary | Résumé

Sets normal priority on bulk send for a low volume threshold. 

Because we recently made a change to set all newly created templates to `bulk` priority (i.e. slow), in order for new templates to be manually promoted instead of constant manual maintenance in order to keep our priority lanes quick, we hit scenarios where low bulk volume (as in test of 1 notification) are put into the bulk send, even if we know this isn't bulk due to the number of rows. 

Hence this PR sets a condition for bulk jobs to be automatically redirected to normal queue if it is below the bulk threshold.

* Replaced the default queue of `None` to `normal-tasks`
* Made the queue choosing functions orthogonal to each other (database vs sending) to know which one serves which purpose.
* There is a scenario where the initial priority is preserved: **if the template was set in high priority and the normal threshold is reached (but not bulk): it will still end up in the high priority queue.**

# Test instructions | Instructions pour tester la modification

* Send bulk notifications through the API, testing edge values (normal threshold: 199, bulk threshold: 200).

# Release Instructions | Instructions pour le déploiement

None.

> Necessary steps to perform before and after the deployment of these changes.
> For example, emptying the cache on a feature that changes the cache data
> structure in Redis could be mentioned.

---

> Étapes nécessaires à exécuter avant et après le déploiement des changements
> introduits par cette proposition. Par exemple, vider la cache suite à des
> changements modifiant une structure de données de la cache pourrait être
> mentionné.

# Reviewer checklist | Liste de vérification du réviseur

This is a suggested checklist of questions reviewers might ask during their
review | Voici une suggestion de liste de vérification comprenant des questions
que les réviseurs pourraient poser pendant leur examen :

- [ ] Is the code maintainable? | Est-ce que le code peut être maintenu?
- [ ] Have you tested it? | L’avez-vous testé?
- [ ] Are there automated tests? | Y a-t-il des tests automatisés?
- [ ] Does this cause automated test coverage to drop? | Est-ce que ça entraîne
      une baisse de la quantité de code couvert par les tests automatisés?
- [ ] Does this break existing functionality? | Est-ce que ça brise une
      fonctionnalité existante?
- [ ] Does this change the privacy policy? | Est-ce que ça entraîne une
      modification de la politique de confidentialité?
- [ ] Does this introduce any security concerns? | Est-ce que ça introduit des
      préoccupations liées à la sécurité?
- [ ] Does this significantly alter performance? | Est-ce que ça modifie de
      façon importante la performance?
- [ ] What is the risk level of using added dependencies? | Quel est le degré de
      risque d’utiliser des dépendances ajoutées?
- [ ] Should any documentation be updated as a result of this? (i.e. README
      setup, etc.) | Faudra-t-il mettre à jour la documentation à la suite de ce
      changement (fichier README, etc.)?

### CODE REVIEW

File: app/user/rest.py [hunk 1]
```diff
@@ -440,7 +440,8 @@ def send_contact_request(user_id):
     try:
         contact = ContactRequest(**request.json)
         user = get_user_by_email(contact.email_address)
-        if not any([not s.restricted for s in user.services]):
+        # If the user has no live services, don't want to escalate the ticket.
+        if all([s.restricted for s in user.services]):
```
Comment (jimleroyer) [inline app/user/rest.py:444]: This is unrelated and actually a simplification suggested by Github copilot.

File: poetry.lock
Comment (jimleroyer) [inline poetry.lock:1]: I regen'ed the lock file as I removed the sentry-sdk dependency.

File: tests/conftest.py [hunk 2]
```diff
@@ -71,7 +71,7 @@ def grant_test_db(writer_uri, uri_db_reader):
     db_reader = urlparse(uri_db_reader).username
     db_reader_password = urlparse(uri_db_reader).password
 
-    postgres_db = sqlalchemy.create_engine(writer_uri, echo=False, isolation_level="AUTOCOMMIT", client_encoding="utf8")
+    postgres_db = sqlalchemy.create_engine(writer_uri, echo=False, isolation_level="SERIALIZABLE", client_encoding="utf8")
```
Comment (jimleroyer) [inline tests/conftest.py:74]: I am not sure about this. I might revert this, especially if devs complain in this PR.

Comment (sastels) [reply tests/conftest.py:74]: Did that help with running the tests?

File: requirements-app.txt [hunk 3]
```diff
@@ -26,7 +26,6 @@ PyJWT==2.6.0
 pytz==2021.3
 PyYAML==5.4.1
 SQLAlchemy==1.4.37
-sentry-sdk[flask]==1.10.1
```
Comment (sastels) [inline requirements-app.txt:29]: pretty sure we can delete this file entirely since we've moved to Poetry, but we can try that in a different PR.

File: app/celery/tasks.py [hunk 4]
```diff
@@ -668,15 +657,55 @@ def process_incomplete_job(job_id):
...
+        # If the size isn't a concern, fall back to the template's process type.
+        if process_type == PRIORITY:
+            return QueueNames.PRIORITY_DATABASE
+        elif process_type == BULK:
+            return QueueNames.BULK_DATABASE
+        else:
+            return QueueNames.NORMAL_DATABASE
+
+
+def choose_sending_queue(process_type: str, notif_type: str, notifications_count: int) -> Optional[str]:
```
Comment (sastels) [inline app/celery/tasks.py:683]: 👍 I like how you've refactored these two methods

File: tests/app/celery/test_tasks.py [hunk 5]
```diff
@@ -861,26 +872,44 @@ def test_process_rows_sends_save_task(
...
+            (2, 1_000, PRIORITY, "priority-tasks"),  # keep priority when no thresholds are met
+            (0, 1, PRIORITY, "bulk-tasks"),  # autoswitch to bulk queue if bulk threshold is met, even if in priority.
+            (1, 1, PRIORITY, "bulk-tasks"),  # autoswitch to bulk queue if bulk threshold is met.
+            (1, 1, NORMAL, "bulk-tasks"),  # autoswitch to bulk queue if bulk threshold is met.
+            (1, 1_000, NORMAL, "send-email-tasks"),  # keep normal priority
+            (2, 1_000, NORMAL, "send-email-tasks"),  # keep normal priority
+            (1_000, 1_000, NORMAL, "send-email-tasks"),  # keep normal priority
+            (1, 1, BULK, "bulk-tasks"),  # keep bulk priority
+            (1, 1_000, BULK, "send-email-tasks"),  # autoswitch to normal queue if normal threshold is met.
+            (2, 1_000, BULK, "bulk-tasks"),  # keep bulk priority
```
Comment (sastels) [inline tests/app/celery/test_tasks.py:895]: So if they're set to `BULK` and send only 1 email (ie less than the normal threshold) they'll stay as bulk? But if they sent more to meet the threshold they'd be normal. That seems odd. Shouldn't all `bulk` csvs that are below the bulk threshold get sent as normal?

**Comment by sastels** [comment]
Tested locally and does what it says! 🎉 
Only concerned / slightly confused that
- we don't allow big PRIORITY jobs
- if a BULK job is **too** small then it won't be normal, it'll stay as bulk

File: requirements-app.txt (hunk 3 above)
Comment (jimleroyer) [reply requirements-app.txt:29]: Hmm yeah, we can try in a different PR to isolate some potential side effect. Let me open a new one exclusively for that.

File: tests/app/celery/test_tasks.py (hunk 5 above)
Comment (jimleroyer) [reply tests/app/celery/test_tasks.py:895]: oh yes you might be right, let me review this

**Comment by jimleroyer** [comment]
Thanks for the feedback @sastels . Regarding your concerns:

> we don't allow big PRIORITY jobs

This behavior did not change, but we can discuss offline. The summary is if we allow big priority jobs, we basically lose the speed of priority and its effectiveness. 

> if a BULK job is too small then it won't be normal, it'll stay as bulk

I'll take a look at this. 👍

File: tests/app/celery/test_tasks.py [hunk 6]
```diff
@@ -861,26 +872,44 @@ def test_process_rows_sends_save_task(
...
+                "queue": "send-{}-tasks".format(template_type),
                 "client_reference": reference,
                 "sender_id": str(sender_id) if sender_id else None,
             }
         )
         task_mock.apply_async.assert_called_once()
 
     @pytest.mark.parametrize(
-        "csv_threshold, expected_queue",
+        "csv_normal_threshold, csv_bulk_threshold, template_process_type, expected_queue",
```
Comment (sastels) [inline tests/app/celery/test_tasks.py:883]: can 🔥 the `csv_normal_threshold` here and in the test

File: tests/conftest.py (hunk 2 above)
Comment (jimleroyer) [reply tests/conftest.py]: I reverted back in the end.

File: tests/app/celery/test_tasks.py (hunk 6 above)
Comment (jimleroyer) [reply tests/app/celery/test_tasks.py:883]: removing...

Comment (jimleroyer) [reply tests/app/celery/test_tasks.py]: This was removed. 👍

File: .devcontainer/devcontainer.json [hunk 7]
```diff
@@ -29,22 +29,25 @@
...
 		"mtxr.sqltools",
 		"mtxr.sqltools-driver-pg",
-		"bungcip.better-toml",
-		"matangover.mypy"
+		"pmbenjamin.vscode-snyk",
+		"usernamehw.errorlens",
+		"visualstudioexptteam.vscodeintellicode",
+		"wenfangdu.jump",
+		"wholroyd.jinja",
+		"yzhang.markdown-all-in-one",
```
[the remark this answers is not in the document]
Comment (jimleroyer) [reply .devcontainer/devcontainer.json:50]: Also, 
```
"GitHub.copilot-labs",
```

**Review by sastels — APPROVED** [review]
Alrighty this should improve things for folks!
